src/Models/UserModel.php

```diff
@@ -185,48 +192,111 @@ public function activate(User $user): void
+        $this->tempUser = $data instanceof User ? $data : null;
 
-            if ($result && $data instanceof User) {
-                /** @var User $user */
-                $user = $data->id === null
-                    ? $this->find($this->db->insertID())
-                    : $data;
+        $result = parent::insert($data, $returnID);
 
-                if (! $user->saveEmailIdentity()) {
-                    throw new RuntimeException('Unable to save email identity.');
-                }
-            }
+        $this->checkQueryReturn($result);
+
+        return $returnID ? $this->insertID : $result;
+    }
 
-            return $result;
+    /**
+     * Override the BaseModel's `update()` method.
+     * If you pass User object, also updates Email Identity.
+     *
+     * @param array|int|string|null $id
+     * @param array|User            $data
+     *
+     * @throws ValidationException
+     */
+    public function update($id = null, $data = null): bool
+    {
+        $this->tempUser = $data instanceof User ? $data : null;
+
+        try {
+            /** @throws DataException */
+            $result = parent::update($id, $data);
         } catch (DataException $e) {
             $messages = [
-                lang('Database.emptyDataset', ['insert']),
                 lang('Database.emptyDataset', ['update']),
             ];
+
             if (in_array($e->getMessage(), $messages, true)) {
-                // @TODO Why true? Shouldn't this workaround be removed?
+                $this->tempUser->saveEmailIdentity();
+
                 return true;
             }
 
             throw $e;
         }
+
+        $this->checkQueryReturn($result);
+
+        return true;
+    }
+
+    /**
+     * Override the BaseModel's `save()` method.
+     * If you pass User object, also updates Email Identity.
+     *
+     * @param array|User $data
+     *
+     * @throws ValidationException
+     */
+    public function save($data): bool
+    {
+        $result = parent::save($data);
+
+        $this->checkQueryReturn($result);
+
+        return true;
+    }
+
+    /**
+     * Save Email Identity
+     *
+     * Model event callback called by `afterInsert` and `afterUpdate`.
+     */
+    protected function saveEmailIdentity(array $data): array
+    {
+        // If insert()/update() gets an array data, do nothing.
+        if ($this->tempUser === null) {
+            return $data;
+        }
+
+        // Insert
+        if ($this->tempUser->id === null) {
+            /** @var User $user */
+            $user = $this->find($this->db->insertID());
+
+            $user->email         = $this->tempUser->email ?? '';
+            $user->password      = $this->tempUser->password ?? '';
+            $user->password_hash = $this->tempUser->password_hash ?? '';
+
+            $user->saveEmailIdentity();
+
+            return $data;
+        }
+
+        // Update
+        $this->tempUser->saveEmailIdentity();
+
```

Comment:
I know `$tempUser` is reset on each call but I would feel better if it were also reset after the identity update, unless there is some logical reason to keep it around. 
```suggestion
            $this->tempUser = $user;
        }

        // Update and reset
        $this->tempUser->saveEmailIdentity();
        $this->tempUser = null;
```

Reply:
Added `$this->tempUser = null`.